### PR TITLE
[AERIE-1384] Multiple simulation workers

### PR DESCRIPTION
**Pull request by patkenneally**
* **Tickets addressed:** AERIE-1384
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This PR introduces a multi-tenancy capability for Aerie simulation. Prior to this Aerie was able to execute a single simulation run at a time, with the merlin-server queuing simulation requests. Aerie is a multi-user application and needs to be able to execute multiple concurrent simulation requests. Additionally, it would be advantageous if users are able to deploy Aerie across a number of hosts, with the majority of containers on modest hardware and the merlin-worker containers on high performance hardware tuned for simulation execution speed. Threading concurrent simulation requests would have been a viable first step on the path if not for the SPICE library's thread safety issues. Given these considerations, we've chosen a multiprocessing focused solution.

A small merlin-worker application is added. The worker subsumes the responsibility of running a simulation and writing the results to the merlin postgres database. The merlin-server processes a simulation request, creates the initial dataset tables and associations. The simulation_dataset entry is initially set with a status "pending". This status indicates that the simulation run is queued waiting for a worker to take the job. When a worker takes the job the status is updated (by the worker) to "incomplete".   

The postgres notification feature is used to NOTIFY workers LISTEN(ing) for newly created simulation jobs. The added database trigger calls a postgres function to emit a notification with a payload containing the simulation revision data, plan id, simulation id, and dataset id. Upon receiving a notification a worker will attempt to claim the simulation job by changing it's simulation_dataset status field to "incomplete". A simulation job is considered taken if this SQL operation is successful. Then the successful worker continues with executing the simulation. If the SQL operation is unsuccessful (e.g. another worker was successful in claiming a job earlier) then the worker handles the failure to claim the job and moves on to attempt claiming other jobs. We can be assured that there will be no race conditions due to postgres' ACID transaction processing. 

Of particular note is the relocation of profile, event, and span partition table creation to the worker. It was discovered that a DB deadlock is induced by making multiple concurrent simulation requests. Creating these partition tables at the moment they are needed, and in a separate transactions prevents the recurrence of the observed deadlock.  

Users configure a fixed number of workers by adding additional worker service entries in their docker compose. The containers are set to automatically restart so that workers return even in the case of process ending failures.

## Verification
A test script which created a single Banananation mission model, and six different plans for the mission model. Six concurrent simulation requests were made (one for each plan). A successful simulation run was observed and simulation results returned for each request. Further manual testing was done to edit a single plan and re-simulate. 

The ability to run these kinds of integrated tests via the is something that Aerie needs to setup and start running as a portion of our devops processes. There are a number of tickets capturing this in the backlog. They will be picked up in the coming sprints.

I'm happy to share my test script and data with anyone who wants/needs to exercise this PR in the same way.

## Documentation
- Deployment documentation
- Release 
	- documentation
	- products

## Future work
None

### CODE REVIEW

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/ResultsProtocol.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/ResultsProtocol.java]: 🔵 We can uncomment `sealed`, since we're on Java 17 now

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateSimulationDatasetAction.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateSimulationDatasetAction.java]: Just making a note that this will need to be carefully rebased onto #102 , since this change was made independently in both PRs.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ClaimSimulationAction.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ClaimSimulationAction.java]: 🔵 Idle thought - since this is expected, rather than exceptional behavior, I wonder if it would make sense to return an `Optional<Long>` or something to indicate that the claim action failed. I have no strong reason to go either way.

If we do decide that an exception is the preferred mechanism to communicate a failed claim, my next thought is whether we want a specific exception for this scenario 🤔

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java
Comment (mattdailis) [inline merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java:33]: 🔶 I'm intrigued by this unwrap - I wonder if it's possible to avoid it. In particular it seems like `DataSource` could be more specific... I'll let you know if I find an alternate way of expressing this.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java
Comment (mattdailis) [inline merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java]: 🔻 Flagging this catch block - we should either rethrow some kind of an exception, or write a comment explaining why we're suppressing it.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/AerieAppDriver.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/AerieAppDriver.java]: 🔶 This change doesn't seem to do anything besides make the variable type explicit - we can probably revert it.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java:11]: 🔵 Just for context - I remember the original intent of doing these in triggers was to encapsulate partition management in the database itself. That said, deadlocks are no fun, and I'm certainly more comfortable with Java code than PL/pgSQL, so I'm okay with handling this in Java

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: 🔶 I'm wondering if we're suppressing any potentially useful exceptions here. We only expect `SQLException` if we fail to claim, right? A `DatabaseException` would indicate a real error that we shouldn't suppress?

EDIT: I just read further down the file and saw that we're converting `SQLExceptions` to `DatabaseExceptions`, in which case I'm still confused why we're catching both in this catch block.

Not requesting any specific change at the moment, just expressing my confusion 👍

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java
Comment (mattdailis) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java:247]: 🔶 Continuing from previous comments - it seems like "failed to create partitions" will be lumped together with "failed to claim a dataset" since both throw indistinguishable `DatabaseException`s.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java
Comment (mattdailis) [inline merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java]: 🔵 Is it okay for there to be no simulation for a plan that does have an entry in `simulation_dataset`? We may be protected from this by some foreign key relationships 🤔 

By "is it okay" I'm wondering if we need to throw some kind of exception... at the same time, it probably doesn't make sense for a random worker to crash... I'm not sure what the best answer is here.

File: docker-compose.yml
Comment (mattdailis) [inline docker-compose.yml]: 🔵 I'm curious about this port - what is it used for? Should we expose it?

**Review by mattdailis — COMMENTED** [review]
Generally looks really good, and I'm super stoked for this new functionality. Most of my comments are me trying to understand our error handling story.

Key:
🔻 Issues to address before merge
 🔶 Requests that should not block merge, but should at least be discussed 
🔵 Recommendations that can be ignored if desired

File: docker-compose.yml
Comment (patkenneally) [reply docker-compose.yml]: Good catch. A rogue copy and paste.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java:11]: Is your implication here that we could be triggering the partition creation on the change of status from "pending" to "incomplete" and thus it could remain a triggered function within our schema definitions?

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/AerieAppDriver.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/AerieAppDriver.java]: Thanks for that catch. Failed to fully clean up after an experiment I backed out of :D

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateSimulationDatasetAction.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateSimulationDatasetAction.java]: Agreed. I expect 102 in sooner so will rebase.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java
Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java:11]: I hadn't thought it that far through - but yes, now that you mention it,  I think that would work. I haven't worked with partitions before, so I'm not really sure what to advise here. Maybe we can consult @kroffo or @Twisol or Tung.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java
Comment (mattdailis) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java:33]: From a cursory look at the types, it seems like your approach is the expected approach 👍  I'll resolve this comment now.

Comment (patkenneally) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java:33]: Bummer. I was hoping you would find something I hadn't found :) . I agree, it feels dirty/wrong.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: Given the discussion with @kroffo and @mattdailis I've adjusted both the `ClaimSimulationAction` and the `PostgresResultsCellrepository` which uses it. The action now only returns an SQLException when an unexpected number of result rows are returned by the db update. An `UnclaimableSimulationException` has been added, although no action is taken because of it so I'm unsure if this is the best remedy here. Still pondering...

Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: I think this solution adequately addresses @kroffo 's concerns about using the `SQLException` to mean two different things :+1:

I would maybe request a javadoc comment on `claimSimulationDataset` and/or `ClaimSimulationAction.apply` explaining what the `UnclaimableSimulationException` means.

Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: Some thoughts that went through my head were that if this method were returning some data that the caller needed, we could return an Optional and feel reasonably certain that the caller will use that information.

In this case, the whole point of `claim` is its side-effect - the caller just needs to check whether it succeeded or not. Declaring the checked exception as part of the method signature forces the caller to handle the failure case.

We could potentially contort the data structures to make `claim` return some data that the caller needed (e.g. caller only has `datasetId`, in order to get the rest of the information they call `claim`. Thus if `claim` returns an `Optional.empty`, the caller cannot erroneously proceed and try to run the simulation), but I'm not sure I like that either.

If our external advisory board of @Twisol or @kroffo (or anyone else who has opinions on when to use exceptions vs when to return a success/failure value) wants to weigh in, feel free :+1:

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/ResultsProtocol.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/ResultsProtocol.java]: Done

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java
Comment (patkenneally) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/ListenSimulationCapability.java]: Have resolved this by the resetting the interruption. Reading in "[Java Concurrency In Practice](https://jcip.net/)", they recommend resetting the interrupt when you are in a context that prevent you from taking action. We have no ability to take action within this listener callback.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java
Comment (patkenneally) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java]: I've removed this block of code as it was not needed. It was left over from earlier prototyping efforts.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java
Comment (patkenneally) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: Good idea to add some Javadocs. I did so.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java
Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateEventPartitionAction.java:11]: If you just want the logic centralized in the database, you can always [`CALL`](https://www.postgresql.org/docs/14/sql-call.html) a stored procedure that does the partitioning.

Whether or not this stored procedure is invoked automatically on some change in the database (as a trigger), or manually by the Java client, is an orthogonal concern. 😄

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java
Comment (Twisol) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: I haven't looked at the code (GitHub is being very unhelpful; it says "these commits no longer exist"), but not being able to claim a simulation just means the worker should go back to the top and wait for another simulation. It probably couldn't claim because another worker got there first.

Based on what Matt described just above, I would argue for an `Optional<>` based approach, where an object permitting write access to the simulation is provided only if the claim was successful. I'm not sure that an exception is really suitable for a case that is not only expected, but will likely happen quite often (N-1 times simultaneously for each simulation, where N is the number of idle workers 😆)

Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java]: At the ResultsCellRepository interface level, that's exactly how it works (I think): 

https://github.com/NASA-AMMOS/aerie/blob/11751aaa624f97fa1fbab5f56a91ee694f12d0b8/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/ResultsCellRepository.java#L11

The implementation of the `PostgresResultCellRepository#claim` converts an `UnclaimableSimulationException` into an Optional:

https://github.com/NASA-AMMOS/aerie/blob/11751aaa624f97fa1fbab5f56a91ee694f12d0b8/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java#L90-L115

I think I was musing about the `ClaimSimulationAction`, which is a `void` method that throws an `UnclaimableSimulationException`:

https://github.com/NASA-AMMOS/aerie/blob/11751aaa624f97fa1fbab5f56a91ee694f12d0b8/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ClaimSimulationAction.java#L24

I guess I'm unsure whether it's appropriate for the `UnclaimableSimulationException` to exist. My current stance is that since its local to the repository, it's fine to throw an exception - but I don't feel like I have a good sense of when it makes sense to use an exception, and when a "failure" should be part of the return type.

**Review by mattdailis — APPROVED** [review]

**Review by adrienmaillard — APPROVED** [review]

**Pull request merged**
